### PR TITLE
fix(16337): close DatePicker calendar on away click when inside Modal or ComposedModal

### LINKED ISSUE #16337: [Bug]: <DatePicker> hosted inside of <Modal> or <ComposedModal> does not close calendar when clicking outside
### Package

@carbon/react

### Browser

Chrome, Firefox

### Package version

v1.56.0

### React version

v18

### Description

When a <DatePicker> is hosted inside of either <Modal> or <ComposedModal>, click on the <DateInput> and the calendar popover can not be dismissed by clicking outside of the calendar.

https://github.com/carbon-design-system/carbon/assets/38996291/6623e372-5dfe-441c-a2c4-028c886c1848

### Reproduction/example

https://stackblitz.com/edit/github-datbmn?file=src%2FApp.jsx

### Steps to reproduce

1. The example already has the modal opened by default. Click on the date input, and no matter where you click inside the `<ComposedModal>` and outside of the calendar, the calendar just does not close. It can only be closed by tab away or select the date.
1. Close the modal by clicking the close button.
2. There is another `<DatePicker` outside of `<ComposedModal>`. It has no problem closes its calendar popover by clicking elsewhere in the page.

### Suggested Severity

None

### Application/PAL

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [X] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

### PR DESCRIPTION
**Pull request by 2nikhiltom**
Closes https://github.com/carbon-design-system/carbon/issues/16337

The root bug with  DatePicker was : When using the datepicker with a range selection, the end date calendar does not close when the user clicks away, causing it to always stay open ([Issue](https://github.com/carbon-design-system/carbon/issues/16740) -> [PR](https://github.com/carbon-design-system/carbon/pull/16742))

Fix for https://github.com/carbon-design-system/carbon/issues/16337 is dependent on  [PR](https://github.com/carbon-design-system/carbon/pull/16742)
Also this issue was happing because `evt.stopPropagation()` was added on `Mousedown` event to add support for [Nest modal](https://github.com/carbon-design-system/carbon/pull/14320)

To fix https://github.com/carbon-design-system/carbon/issues/16337  : I thought to keep this action behind a prop 

- Introduced a new prop `isNested`, in modal.tsx.
- The event handler will now run `evt.stopPropagation()` only when the `isNested` prop is present.

image

- Additionally, we can update the documentation to include guidelines advising against the use of DatePickers within nested modals.

I am looking for reviews and suggestions on this.
If `isNested` prop works as a fix , I will be implementing this is in `ComposedModal` as well 

#### Testing / Reviewing

- Take pull in local
- Open Modal storybook->default 
- Select start and end date in the datepicker 
- Open end date calendar, do not select any date and click outside of calendar and verify if calendar closes 

Verify the functionlity when `isNested`  is passed 
- Pass `isNested` as a prop in `Modal`
- Select start and end date in the datepicker 
- 
image

-  Open end date calendar, do not select any date and click outside of calendar and verify that does not close

### CODE REVIEW

File: packages/react/src/components/DatePicker/plugins/fixEventsPlugin.js
Comment (2nikhiltom) [inline packages/react/src/components/DatePicker/plugins/fixEventsPlugin.js:1]: The changes in fixEventsPlugin.js will be made available from https://github.com/carbon-design-system/carbon/pull/16742

File: packages/react/src/components/Modal/Modal.stories.js [hunk 1]
```diff
@@ -45,6 +47,7 @@ export const Default = () => {
         modalHeading="Add a custom domain"
         modalLabel="Account resources"
         primaryButtonText="Add"
+        //isNested
```
Comment (riddhybansal) [inline packages/react/src/components/Modal/Modal.stories.js:50]: I think this comment can be removed!!

File: packages/react/src/components/DatePicker/plugins/fixEventsPlugin.js [hunk 2]
```diff
@@ -109,9 +109,10 @@ export default (config) => (fp) => {
         if (inputTo.value) {
           fp.setDate(
             [inputFrom.value, inputTo.value],
-            true,
+            false,
```
Comment (riddhybansal) [inline packages/react/src/components/DatePicker/plugins/fixEventsPlugin.js:112]: Does previously sending the second value as true , was forcefully firing onChange event ?

**Review by riddhybansal — COMMENTED** [review]
This looks like a great hack @2nikhiltom . Great work I must say !!

**Pull request closed without merging**